hawc/apps/invitro/exports.py

```diff
@@ -1,362 +1,622 @@
+            ),
+        }
+
+
+class IVBenchmarkExport(ModelExport):
+    def get_value_map(self):
+        return {
+            "id": "id",
+            "benchmark": "benchmark",
+            "value": "value",
+        }
+
+
+class InvitroExporter(Exporter):
+    def build_modules(self) -> list[ModelExport]:
+        return [
+            StudyExport(
+                "study",
+                "experiment__study",
+                include=(
+                    "id",
+                    "hero_id",
+                    "pubmed_id",
+                    "doi",
+                    "short_citation",
+                    "study_identifier",
+                    "published",
+                ),
+            ),
+            IVChemicalExport(
+                "iv_chemical",
+                "chemical",
+            ),
+            DSSToxExport(
+                "dsstox",
+                "chemical__dtxsid",
+            ),
+            IVExperimentExport(
+                "iv_experiment",
+                "experiment",
+            ),
+            IVCellTypeExport(
+                "iv_cell_type",
+                "experiment__cell_type",
+            ),
+            IVEndpointExport(
+                "iv_endpoint",
+                "",
+                exclude=(
+                    "data_type",
+                    "variance_type",
+                ),
+            ),
+            IVEndpointGroupExport(
+                "iv_endpoint_group",
+                "groups",
+                include=(
+                    "id",
+                    "dose",
+                    "difference_control_display",
+                    "significant_control",
+                    "cytotoxicity_observed",
+                ),
+            ),
+            IVBenchmarkExport(
+                "iv_benchmark",
+                "benchmarks",
+            ),
+        ]
 
 
 class DataPivotEndpoint(FlatFileExporter):
-    def _get_header_row(self):
-        if self.queryset.first() is None:
-            self.rob_headers, self.rob_data = {}, {}
-        else:
-            study_ids = set(self.queryset.values_list("experiment__study_id", flat=True))
-            self.rob_headers, self.rob_data = FinalRiskOfBiasScore.get_dp_export(
-                self.queryset.first().assessment_id,
-                study_ids,
-                "invitro",
+    def handle_dsstox(self, df: pd.DataFrame) -> pd.DataFrame:
+        # condenses the dsstox info into one column
+        dsstox_cols = [col for col in df.columns if col.startswith("dsstox-")]
+        dsstox_df = df[dsstox_cols]
+        dsstox_df.columns = dsstox_df.columns.str[7:]
+        df["chemical DTXSID"] = dsstox_df.to_dict(orient="records")
+        df["chemical DTXSID"] = df["chemical DTXSID"].apply(
+            lambda x: None if pd.isna(x["dtxsid"]) else x
+        )
+        return df.drop(columns=dsstox_cols)
+
+    def handle_dose_groups(self, df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
wow, this is impressive, and really complicated. I think keeping it as is is good for now, this seems like a conservative conversion from our initial approach. However, when I try it on a large dataset, it's slow. Not super slow, so it may be good enough.

I'll make a card in our backlog to explore more aggressive optimizations. It'll be easier to explore the impact of the optimizations once we have the jupyter notebooks you'll be building for comparing different approaches.

To document some exploration ideas, I got pretty far with this which ran really quickly:

```python
df = pd.DataFrame(IVEndpointGroup.objects.filter(endpoint__assessment=123).order_by('endpoint', 'dose_group_id').values())
col_vals = ['dose','n','response','variance','difference_control','significant_control','cytotoxicity_observed']
df1 = pd.pivot(df.query('dose > 0'), values=col_vals,index=['endpoint_id'],columns=['dose_group_id'])
df1.columns = df1.columns.map(lambda el: f'{el[0].replace("_", " ").title()} {el[1]+1}')
df1.fillna("-")
```